File: app/src/lib/game/computer/modules/card-probability.ts
import { SUIT_VALUES } from '../../types.js';
import type { Candidate, ModuleContext, ScoringModule } from '../types.js';

const MAX_PER_SHAPE = SUIT_VALUES.length; // 12

// Prefer landing the discard pile on a shape that has fewer cards remaining
// in the system (discard + own hand already subtracted in shapeRemaining).
// Fewer remaining → opponents less likely to hold that shape → harder for them to play.
// Kept as a tiebreaker (max 2) so it nudges but never overrides action-awareness or thinning.
export const cardProbability: ScoringModule = (candidate: Candidate, ctx: ModuleContext): number => {
	if (candidate.kind !== 'play-suit') return 0;

	const remaining = ctx.shapeRemaining[candidate.card.shape];
	const scale = 1 / Math.max(1, ctx.state.players.length - 1);
	return ((MAX_PER_SHAPE - remaining) / MAX_PER_SHAPE) * 2 * scale;
};

File: server/src/game/ai/modules/card_probability.rs
use crate::game::{
    ai::types::{Candidate, ModuleContext},
    engine::CARDS_PER_SHAPE,
};

/// Prefer landing on a shape that has fewer cards remaining — harder for opponents to match.
pub fn card_probability(candidate: &Candidate, ctx: &ModuleContext<'_>) -> f64 {
    let Candidate::PlaySuit { shape, .. } = candidate else {
        return 0.0;
    };

    let remaining = *ctx.shape_remaining.get(shape).unwrap_or(&0) as f64;
    let scale = 1.0 / (ctx.state.seats.len() as f64 - 1.0).max(1.0);
    ((CARDS_PER_SHAPE as f64 - remaining) / CARDS_PER_SHAPE as f64) * 2.0 * scale
}
